### DISCUSSION TITLE
Support Hyper for bindings

### DISCUSSION THREAD

**Discussion by morenathan** [niri-wm/niri#312] [Ideas]
I see where Shift, Control, Super, and Alt are supported. Mod5/altgr is support as well; But I don't see where Hyper can be bound. The documentation doesn't seem to allow me to bind based on any custom key either.

This would be wonderful to also have available and is traditional to have on a Unix keyboard.

**Comment by YaLTeR** [comment]
Is that another xkb modifier? If so, then it needs to be added to the Smithay enum first. 

Also, I thought generally what people call Hyper is a combination of several normal modifiers? If so then you can bind that by spelling them out.

**Reply by morenathan** [reply]
No, it's an actual key. Windows has typically called "Hyper" a combination of Ctrl+Alt+Shift, but under xkb it's generally Mod4 under X11 iirc.

I'm not too familiar with any of these new C variants, but I can look there :)

**Reply by YaLTeR** [reply]
You can check this as an example: https://github.com/Smithay/smithay/pull/1365

**Reply by YaLTeR** [reply]
And this: https://github.com/YaLTeR/niri/pull/261

**Reply by morenathan** [reply]
from xkb/keycodes:
```
    <LVL5> =     8;
    alias <MDSW> = <LVL5>;
    <LVL3> =   124; // <U7C>
    <ALT>  =   125; // <U7D>
    <META> =   156; // <I1C>
    <SUPR> =   127; // <U7F>
    <HYPR> =   128; // <U80>
```

**Reply by wizetek** [reply]
+1 to have this ability in niri. Caps Lock with its original functionality is useless to me. I like to remap it to Hyper (not Super) which gives me a nice new large set of key combos. I already use Alt_R (AltGr) [Mod5] for international characters.

As far as I understand from xkb perspective both Super and Hyper are Mod4. But they can be separated.

My config input section looks like this:
````
input {
    keyboard {
        xkb {
            layout "pl,us"
            options "grp:shifts_toggle,caps:hyper"
````
I would bind as follows – Hyper should suffice but it can be more specific if there are two of them: L/R
````
binds {
    Hyper+T { spawn "terminator"; }                                                               
    //Hyper_L+T { spawn "terminator"; }
````
This is currently not valid though.

Meanwhile `wev` happily produces
````
[14:     wl_keyboard] key: serial: 838; time: 117996790; key: 66; state: 1 (pressed)
                      sym: Hyper_L      (65517), utf8: ''
[14:     wl_keyboard] modifiers: serial: 0; group: 0
                      depressed: 00000020: Mod3
                      latched: 00000000
                      locked: 00000000
[14:     wl_keyboard] key: serial: 839; time: 117996862; key: 66; state: 0 (released)
                      sym: Hyper_L      (65517), utf8: ''
[14:     wl_keyboard] modifiers: serial: 0; group: 0
                      depressed: 00000000
                      latched: 00000000
                      locked: 00000000
````
Is this a feature request for niri or smithay first?

**Reply by YaLTeR** [reply]
Smithay I guess

**Reply by hpr1999** [reply]
It's been a while, but I wanted to give this a humble +1.

Elsewhere, Hyper is also often defined as Super, Ctrl, Alt and Shift combined. I can use this fine with my qmk enabled keyboard, it's just a little cumbersome to type all those modifiers in the config.

However, localectl also has the useful option of mapping caps lock to left control and left control to hyper. I was hoping to use this as part of my mappings for my laptop, but alas, no luck, since this outputs the HyperL keycode as described above.

**Comment by wizetek** [comment]
Nevermind! This is actually already working. I missed the bit showing "Mod3" in `wev` output.

So just bind like so and you are good to go:
```
binds {
    Mod3+T { spawn "terminator"; }  
```
Maybe in the future niri can also understand Hyper similarly to how Super and Mod are related.

**Reply by YaLTeR** [reply]
The alternative name for Mod3 is ISO_Level5_Shift. I'm not sure where Hyper comes in here

**Reply by wizetek** [reply]
Yeah, somehow when Lock gets remapped to Hyper, it gets assigned to Mod3. Then effectively "lock" is blanked out and Hyper_L is removed from Mod4 leaving only Super (L/R) there. I see the same under X11 with xmodmap.

This is exactly what I was looking for and is consistent with X11, so... case closed (for me at least). Thanks!
